### DISCUSSION TITLE
Feature Request: impl<'a, Parent, Child> BelongingToDsl<(&'a [Parent], &'a [Parent])> for Child

### DISCUSSION THREAD

**Discussion by Bajix** [diesel-rs/diesel#3559] [Ideas]
I authored [stack-queue](https://crates.io/crates/stack-queue) specifically so that I could create a data loader that integrates with Diesel. The current design only batches contiguous slices and simply starts a new batch whenever a task batch would otherwise wrap around to the beginning, however I could make this support wrap-around batching by returning an ordered pair of slices instead. The only way I could make this work and still integrate with BelongingToDsl without creating an intermediary Vec would be for BelongingToDsl to impl<'a, Parent, Child> BelongingToDsl<(&'a [Parent], &'a [Parent])> for Child

Alternatively this could work with an iterator over &Parent

**Comment by weiznich** [comment]
I'm happy to accept a PR for the iterator solution, as long as it does not break existing usages. That written: My gut feeling tells me that this is likely not possible due to conflicting trait impls.

**Reply by Bajix** [reply]
I had trait conflicts with doing this via iterators but not such issue with a tuple of slices

```
error[E0119]: conflicting implementations of trait `belonging_to_dsl::BelongingToDsl<&_>`
   --> diesel/src/associations/belongs_to.rs:169:1
    |
136 | impl<'a, Parent, Child> BelongingToDsl<&'a Parent> for Child
    | ------------------------------------------------------------ first implementation here
...
169 | impl<'a, Parent, Child, Iter> BelongingToDsl<Iter> for Child
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation
    |
    = note: downstream crates may implement trait `associations::Identifiable` for type `&_`
    = note: downstream crates may implement trait `std::iter::Iterator` for type `&_`
    = note: upstream crates may add a new impl of trait `std::iter::Iterator` for type `std::vec::Vec<_>` in future versions
```


```rust
impl<'a, Parent, Child, Iter> BelongingToDsl<Iter> for Child
where
    Iter: IntoIterator<Item = &'a Parent>,
    &'a Parent: Identifiable,
    Child: HasTable + BelongsTo<Parent>,
    Vec<Id<&'a Parent>>: AsInExpression<<Child::ForeignKeyColumn as Expression>::SqlType>,
    <Child as HasTable>::Table: FilterDsl<EqAny<Child::ForeignKeyColumn, Vec<Id<&'a Parent>>>>,
    Child::ForeignKeyColumn: ExpressionMethods,
    <Child::ForeignKeyColumn as Expression>::SqlType: SqlType,
{
    type Output = Filter<Child::Table, EqAny<Child::ForeignKeyColumn, Vec<Id<&'a Parent>>>>;

    fn belonging_to(parents: Iter) -> Self::Output {
        let ids = parents.map(Identifiable::id).collect::<Vec<_>>();
        FilterDsl::filter(Child::table(), Child::foreign_key_column().eq_any(ids))
    }
}
```

**Comment by Bajix** [comment]
Added PR for tuple of slices impl:
https://github.com/diesel-rs/diesel/pull/3561
